src/video/psp/SDL_pspmessagebox.c

```diff
@@ -0,0 +1,118 @@
+/*
+  Simple DirectMedia Layer
+  Copyright (C) 1997-2024 Sam Lantinga <slouken@libsdl.org>
+
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+
+#include "SDL_internal.h"
+
+#ifdef SDL_VIDEO_DRIVER_PSP
+
+#include "SDL_pspvideo.h"
+#include "SDL_pspmessagebox.h"
+#include <psputility.h>
+#include <pspgu.h>
+#include <pspdisplay.h>
+#include <pspthreadman.h>
+#include <pspkernel.h>
+
+static void configure_dialog(pspUtilityMsgDialogParams *dialog, size_t dialog_size)
+{
+	/* clear structure and setup size */
+	SDL_memset(dialog, 0, dialog_size);
+	dialog->base.size = dialog_size;
+
+	/* set language */
+	sceUtilityGetSystemParamInt(PSP_SYSTEMPARAM_ID_INT_LANGUAGE, &dialog->base.language);
+
+	/* set X/O swap */
+	sceUtilityGetSystemParamInt(PSP_SYSTEMPARAM_ID_INT_UNKNOWN, &dialog->base.buttonSwap);
+
+	/* set thread priorities */
+	/* TODO: understand how these work */
+	dialog->base.soundThread = 0x10;
+	dialog->base.graphicsThread = 0x11;
+	dialog->base.fontThread = 0x12;
+	dialog->base.accessThread = 0x13;
+}
+
+int PSP_ShowMessageBox(const SDL_MessageBoxData *messageboxdata, int *buttonID)
+{
+	pspUtilityMsgDialogParams dialog;
+	int status;
+
+	/* TODO: AGHHHHHhhhHH */
+	if (SDL_WasInit(SDL_INIT_VIDEO) == 0 || SDL_GL_GetCurrentContext() == NULL || SDL_GL_GetCurrentWindow() == NULL)
```

Comment:
Are these really a problem? Does sceUtilityMsgDialogInitStart() fail if an OpenGL context isn't current?

Reply:
No idea about that function specifically, but it completely fails to function properly if there's no active GL (aka `sceGu`) context running at the time. If you call this function before creating a window & GL context, it will lead to either a blank screen or a heavily flickering screen (and no dialog) that persists after the user application is initialized.

I've been working on this problem with my PSP for the last 2 hours and I haven't figured out the appropriate way to create a new GL context to display the dialog, and then tear it down in a way that doesn't lead to strange behaviour in the user application.

Reply:
Having  `SDL_WasInit(SDL_INIT_VIDEO) == 0` here is good, but because `SDL_GL_GetCurrentContext` and `SDL_GL_GetCurrentWindow` do not work if OpenGL is not used they should be removed.

What would be best is to have some check if there is a renderer at all, but we don't have access to this at this level. We could look in the SDL_MessageBoxData to get the renderer from the window, but that's an optional parameter, so it's not reliable.

When there is no renderer, the message box will trigger and work, but it won't be visible, which is confusing. I think it would be acceptable to keep that problem in there if there is no other way to determine if a render has been created already. I'll do some testing to see if I can find something in libgu.